schematools/validation.py

```diff
@@ -0,0 +1,149 @@
+from schematools.types import DatasetSchema
+from schematools.utils import to_snake_case
+
+
+@dataclass(frozen=True)
+class ValidationError:
+    """Capture validation errors."""
+
+    validator_name: str
+    message: str
+
+    def __str__(self) -> str:
+        return f"[{self.validator_name}] {self.message}"
+
+
+class ValidationException(Exception):
+    """Raised when validation fails to execute.
+
+    .. note::
+
+       This is not for validation errors. See :class:`ValidationError` instead.
+    """
+
+    message: str
+
+    def __init__(self, message: str) -> None:
+        super().__init__(*message)
+        self.message = message
+
+
+class Validator:
+    """Base class for validators.
+
+    Not only is this a base class for validators,
+    it is also used for running all registered validators.
+    See Also: :meth:`run_all`
+
+    Registration is a side-effect of overriding this base class.
+
+    """
+
+    _registry: ClassVar[List[Type[Validator]]] = []
+    dataset: DatasetSchema
+
+    @classmethod
+    def __init_subclass__(cls: Type[Validator]) -> None:
+        """Register sub classes."""
+        super().__init_subclass__()
+        cls._registry.append(cls)
+
+    def __init__(self, dataset: DatasetSchema) -> None:
+        """Initialize the validator with a dataset.
+
+        Args:
+            dataset: The dataset to run the validations on.
+        """
+        self.dataset = dataset
+
+    def validate(self) -> Iterator[ValidationError]:
+        """Run validation."""
+        raise NotImplementedError(
+            f"{self.__class__.__name__}.{self.validate.__name__} should be overridden in "
+            f"subclasses and called from there."
+        )
+
+    @final
+    def run_all(self) -> Iterator[ValidationError]:
+        """Run all registered validators.
+
+        Yields:
+            :class:`ValidationError`s if any.
+
+        """
+        for validator_cls in self._registry:
+            validator_inst = validator_cls(dataset=self.dataset)
+            yield from validator_inst.validate()
+
+
+class PsqlIdentifierLengthValidator(Validator):
+    """Validate inferred PostgreSQL table names for not exceeding max length.
+
+    PostgreSQL has a maximum length for identifiers such as table names.
+    We infer table names from dataset and table ids in the schemas.
+    Those inferred table names should not exceed the max identifier length
+    supported by PostgreSQL.
+    """
+
+    def validate(self) -> Iterator[ValidationError]:
+        for table in self.dataset.get_tables(include_nested=True, include_through=True):
+            # `table_name` should probably be a property on `DatasetTableSchema`.
+            # There already is a `db_name` property however it truncates the inferred
+            # table name to prevent exceeding the PostgreSQL limit. Hence it cannot
+            # be used for validation purposes.
+            table_name = to_snake_case("_".join((self.dataset.id, table.id)))
```

Comment:
In another PR I am adding a `name` property to the `DatasetTableSchema`. The `id` is provided by the amsterdam schema. The `name` is usually the same, but can be overridden in the amsterdam schema by a `shortname` property.
What this validation implementations does not cover yet are the tablenames that are generated on the fly for the through tables. That's for a new validator.

Reply:
I thought I would get those through tables automatically by calling:

```python
get_tables(include_nested=True, include_through=True)
```

Are you saying `include_through=True` doesn't include through tables?

Reply:
Will merge the PR. We can always extend the validator at a later time or write a new one as you suggested.

Reply:
:facepalm: I totally missed the `include_through`. I'm glad this usecase is already covered!